src/VisualStudio/Core/Test.Next/Services/ServiceHubServicesTests.cs

```diff
@@ -684,6 +682,698 @@ public async Task InProcAndRemoteWorkspaceAgree20()
                 ImmutableArray.Create(("SG.cs", CreateStreamText(contents, useBOM: useBOM2, useMemoryStream: useMemoryStream2))));
         }
 
+        [PartNotDiscoverable]
+        [ExportWorkspaceService(typeof(IWorkspaceConfigurationService), ServiceLayer.Test), System.Composition.Shared]
+        [method: ImportingConstructor]
+        [method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        private sealed class TestWorkspaceConfigurationService(IGlobalOptionService globalOptionService) : IWorkspaceConfigurationService
+        {
+            public WorkspaceConfigurationOptions Options => globalOptionService.GetWorkspaceConfigurationOptions();
+        }
+
+        [Theory, CombinatorialData]
+        internal async Task TestSourceGenerationExecution_RegenerateOnEdit(
+            SourceGeneratorExecutionPreference executionPreference)
+        {
+            using var workspace = CreateWorkspace([typeof(TestWorkspaceConfigurationService)]);
+
+            var globalOptionService = workspace.ExportProvider.GetExportedValue<IGlobalOptionService>();
+            globalOptionService.SetGlobalOption(WorkspaceConfigurationOptionsStorage.SourceGeneratorExecution, executionPreference);
+
+            var callCount = 0;
+            var generator = new CallbackGenerator(
+                onInit: _ => { },
+                onExecute: _ => { },
+                computeSourceTexts: () =>
+                {
+                    callCount++;
+                    return ImmutableArray.Create(("hint", SourceText.From($"// generated document {callCount}", Encoding.UTF8)));
+                });
+
+            // want to access the true workspace solution (which will be a fork of the solution we're producing here).
+            DocumentId tempDocId;
+            {
+                var projectId = ProjectId.CreateNewId();
+                var analyzerReference = new TestGeneratorReference(generator);
+                var project = workspace.CurrentSolution
+                    .AddProject(ProjectInfo.Create(projectId, VersionStamp.Default, name: "Test", assemblyName: "Test", language: LanguageNames.CSharp))
+                    .GetRequiredProject(projectId)
+                    .AddAnalyzerReference(analyzerReference);
+                var tempDoc = project.AddDocument("X.cs", SourceText.From("// "));
+                tempDocId = tempDoc.Id;
+
+                Assert.True(workspace.SetCurrentSolution(_ => tempDoc.Project.Solution, WorkspaceChangeKind.SolutionChanged));
+            }
```

Comment:
Any reason this isn't a function since it's duplicated a couple times and adds scoping to prevent it's data from being used elsewhere

Reply:
done.